core/src/main/scala/org/apache/spark/scheduler/cluster/YarnSchedulerBackend.scala

```diff
@@ -91,6 +94,36 @@ private[spark] abstract class YarnSchedulerBackend(
   }
 
   /**
+   * Override the DriverEndpoint to add extra logic for the case when an executor is disconnected.
+   * We should check the cluster manager and find if the loss of the executor was caused by YARN
+   * force killing it due to preemption.
+   */
+  private class YarnDriverEndpoint(rpcEnv: RpcEnv, sparkProperties: Seq[(String, String)])
+      extends DriverEndpoint(rpcEnv, sparkProperties) {
```

Comment:
The superclass, CoarseGrainedSchedulerBackend, will use the driver endpoint for messaging as well, right? So then the CoarseGrainedSchedulerBackend's driver endpoint will now handle both calls for RequestExecutors, KillExecutors, etc. - while it also has to handle CoarseGrainedSchedulerBackend.DriverEndpoint's messages as well. I think the separation of concerns of what endpoints process what messages makes it sensible to have the two endpoints.